core/src/main/scala/org/apache/spark/NewAccumulator.scala

```diff
@@ -0,0 +1,299 @@
+   * of overwriting it. This happens when we copy accumulators, e.g. when we reconstruct
+   * [[org.apache.spark.executor.TaskMetrics]] from accumulator updates.
+   */
+  def register(a: NewAccumulator[_, _]): Unit = synchronized {
+    if (!originals.containsKey(a.id)) {
+      originals.put(a.id, new jl.ref.WeakReference[NewAccumulator[_, _]](a))
+    }
+  }
+
+  /**
+   * Unregister the [[NewAccumulator]] with the given ID, if any.
+   */
+  def remove(id: Long): Unit = synchronized {
+    originals.remove(id)
+  }
+
+  /**
+   * Return the [[NewAccumulator]] registered with the given ID, if any.
+   */
+  def get(id: Long): Option[NewAccumulator[_, _]] = synchronized {
+    Option(originals.get(id)).map { ref =>
+      // Since we are storing weak references, we must check whether the underlying data is valid.
+      val acc = ref.get
+      if (acc eq null) {
+        throw new IllegalAccessError(s"Attempted to access garbage collected accumulator $id")
+      }
+      acc
+    }
+  }
+
+  /**
+   * Clear all registered [[NewAccumulator]]s. For testing only.
+   */
+  def clear(): Unit = synchronized {
+    originals.clear()
+  }
+}
+
+
+class IntAccumulator extends NewAccumulator[jl.Integer, jl.Integer] {
+  @transient private[this] var _sum = 0
+
+  override def add(v: jl.Integer): Unit = {
+    _sum += v
+  }
+
+  override def merge(other: NewAccumulator[jl.Integer, jl.Integer]): Unit = other match {
+    case o: IntAccumulator => _sum += o.sum
+    case _ => throw new UnsupportedOperationException(
+      s"Cannot merge ${this.getClass.getName} with ${other.getClass.getName}")
+  }
+
+  override def value: jl.Integer = _sum
+
+  def sum: Int = _sum
+
+  def setValue(newValue: Int): Unit = _sum = newValue
+}
+
+
+class LongAccumulator extends NewAccumulator[jl.Long, jl.Long] {
+  @transient private[this] var _sum = 0L
+
+  override def add(v: jl.Long): Unit = {
+    _sum += v
+  }
+
+  override def merge(other: NewAccumulator[jl.Long, jl.Long]): Unit = other match {
+    case o: LongAccumulator => _sum += o.sum
+    case _ => throw new UnsupportedOperationException(
+      s"Cannot merge ${this.getClass.getName} with ${other.getClass.getName}")
+  }
+
+  override def value: jl.Long = _sum
+
+  def sum: Long = _sum
+
+  def setValue(newValue: Long): Unit = _sum = newValue
+}
+
+
+class DoubleAccumulator extends NewAccumulator[jl.Double, jl.Double] {
+  @transient private[this] var _sum = 0.0
+
+  override def add(v: jl.Double): Unit = {
+    _sum += v
+  }
+
+  override def merge(other: NewAccumulator[jl.Double, jl.Double]): Unit = other match {
+    case o: DoubleAccumulator => _sum += o.sum
+    case _ => throw new UnsupportedOperationException(
+      s"Cannot merge ${this.getClass.getName} with ${other.getClass.getName}")
+  }
+
+  override def value: jl.Double = _sum
+
+  def sum: Double = _sum
+
+  def setValue(newValue: Double): Unit = _sum = newValue
+}
+
+
+class AverageAccumulator extends NewAccumulator[jl.Double, jl.Double] {
```

Comment:
I think getting rid of `setValue` is great, in the consistent accumulators based under the old API I had to just throw an exception if people were using `setValue`